refactor: changed childState to routedState. SirenLink and SubEntity use mixin Routable

tenbradley opened the pull request:
Context:
Basically, we made a routable mixin and this is how it has to be used. I have also made the other Siren classes that should use this mixin use it.

This should make testing of some stuff nicer and routable stuff be more consistent.
Technically part of US121372, but more preventing tech debt for later.

Quality:
- SirenLink and SirenSubEntity pass all tests they previously passed.

tenbradley commented on state/observable/Routable.js:
```diff
@@ -9,12 +11,12 @@ export const Routable = superclass => class extends superclass {
 
-	get childState() {
-		return this._childState;
+	get routedState() {
```
I could be convinced this is a bad name, but it seems more accurate than childState.

AKobets commented on state/observable/Routable.js:
```diff
@@ -32,6 +34,10 @@ export const Routable = superclass => class extends superclass {
 		}
 	}
 
+	createRoutedState(entityID, token) {
```
I could not find where `createRoutedState` is used. Did I miss it?

tenbradley replied:
No, I forgot to change a function call which I have now done.
My plan is to use the term routedState instead of childState as I think it better describes what the state is stored in the routable is.

m6jones commented on state/observable/SirenSubEntity.js:
```diff
@@ -85,15 +63,8 @@ export class SirenSubEntity extends Observable {
-			fetch(this._childState);
-		}
+		this.updateRoutedState(entityId, subEntity);
+		this.childState.setSirenEntity(subEntity);
```
This call should be in updateRoutedState

m6jones replied:
Link doesn't need to do it that is why. But a link will have the href and will be skipped anyways. So it is all good to put it here.

tenbradley replied:
I'm probably going to remove the updateRoutedState function because something about it is causing issues, so this is a moot point.

m6jones commented on state/observable/Routable.js:
```diff
@@ -40,14 +46,25 @@ export const Routable = superclass => class extends superclass {
 
+	async updateRoutedState(identifier, subEntity) {
+		if (this._token) {
+			this.routedState = await this.createRoutedState(identifier, shouldAttachToken(this._token.rawToken, subEntity));
```
Should be below this line. Hmm, createRoutedState could handle it?

m6jones replied:
this = `routedState.setSirenEntity`
setSirenEntity needs to be called before fetch. Otherwise we get a redundant fetch.

tenbradley replied:
I think this is the same question: do we always want to set the routedState's sirenEntity? If so, will do.

tenbradley replied:
Same comment as before, I think I am going to remove this function.

m6jones commented:
One of the tests are failing because in the `HypermediaState.push` accesses childState which is now renamed.

tenbradley commented:
> One of the tests are failing because in the `HypermediaState.push` accesses childState which is now renamed.

Zina and I found that and fixed it and the test was still failing.

tenbradley commented on test/state/HypermediaState.test.js:
```diff
@@ -393,18 +393,18 @@ describe('HypermediaState class', () => {
 			getActionSpy = sinon.spy(getActionObservable);
-			subEntityChilStateSpy = sinon.spy(subEntityObservable.childState);
-			linkPrimeChildStateSpy = sinon.spy(linkPrimeObservable.childState);
+			subEntityRoutedStateSpy = sinon.spy(subEntityObservable.routedState);
+			linkPrimeRoutedStateSpy = sinon.spy(linkPrimeObservable.routedState);
```
Changing childState to routedState fixed two of the tests that were broken.

m6jones commented on state/observable/SirenLink.js:
```diff
@@ -1,13 +1,11 @@
+export class SirenLink extends Routable(Observable) {
 	constructor({ id, token, state } = {}) {
-		super();
+		super({});
```
Should we be adding  token, id and route here?

tenbradley replied:
To the super call? It shouldn't be needed. Neither Routable nor Observable holds any of that data at the moment.

m6jones replied:
Oh,

tenbradley replied:
Oh?

m6jones commented on state/observable/Routable.js:
```diff
@@ -9,12 +11,12 @@ export const Routable = superclass => class extends superclass {
 		this._routes = new Map();
```
This is more for the parameters.

You should give the constructor the parameter, `object = {}` and then do `super(object)` since right now there too much knowledge bleed on one use case for this constructor.

m6jones commented on state/observable/Routable.js:
```diff
@@ -1,3 +1,5 @@
+import { shouldAttachToken } from '../token.js';
```
Not importing fetch
